ethcore/trace/src/types/trace.rs

```diff
@@ -33,6 +34,57 @@ pub struct CallResult {
 	pub output: Bytes,
 }
 
+/// `Call` type. Distinguish between different types of contract interactions.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum CallType {
+	/// Call
+	Call,
+	/// Call code
+	CallCode,
+	/// Delegate call
+	DelegateCall,
+	/// Static call
+	StaticCall,
+}
+
+impl TryFrom<ActionType> for CallType {
+	type Error = &'static str;
+	fn try_from(action_type: ActionType) -> Result<Self, Self::Error> {
+		match action_type {
+			ActionType::Call => Ok(CallType::Call),
+			ActionType::CallCode => Ok(CallType::CallCode),
+			ActionType::DelegateCall => Ok(CallType::DelegateCall),
+			ActionType::StaticCall => Ok(CallType::StaticCall),
+			ActionType::Create => Err("Create cannot be converted to CallType"),
+			ActionType::Create2 => Err("Create2 cannot be converted to CallType"),
+		}
+	}
+}
+
+impl Encodable for CallType {
+	fn rlp_append(&self, s: &mut RlpStream) {
+		let v = match *self {
+			CallType::Call => 0u32,
+			CallType::CallCode => 1,
+			CallType::DelegateCall => 2,
+			CallType::StaticCall => 3,
+		};
+		Encodable::rlp_append(&v, s);
+	}
+}
+
+impl Decodable for CallType {
+	fn decode(rlp: &Rlp) -> Result<Self, DecoderError> {
+		rlp.as_val().and_then(|v| Ok(match v {
+			0u32 => CallType::Call,
```

Comment:
Is this correct? In the old version, when we serialized a 0 this meant CallType::None, but it will now be read to CallType::Call. Shouldn't this all be shifted by 1?

Reply:
`Option<CallType>` is serialized as list https://github.com/paritytech/parity-common/blob/a8329435e19d46b3e50145a9b5a399bfac52a31f/rlp/src/impls.rs#L70-L99 and I assume that #11311 was reviewed properly, so we don't have to re-review it again :sweat_smile:

Reply:
Actually, I think you're right, let me test this again.